### PR TITLE
feat: add CPE to component

### LINKED ISSUE #139: [FEATURE] Add possibility to use CPE in components
The CPE (https://nvd.nist.gov/products/cpe) is still often used to identify vulnerabilities in software and also used by Dependency Track to identify vulnerabilities.

I am currently generating a cyclonedx sbom file with this library with input from various sources where the sources sometimes include the CPE value for the component. The sbom then is imported into DT and other tools to identify vulnerabilities.

Currently I cannot add the CPE to the sbom using this library as the CPE is not supported in the component. 

Pull request for this feature already added.

### PR DESCRIPTION
**Pull request by jblu42**
Setting CPE was missing for component, now it is possible to set CPE and output CPE for a component.

fixes #139

### CODE REVIEW

File: cyclonedx/model/component.py [hunk 1]
```diff
@@ -103,7 +103,7 @@ def __init__(self, name: str, component_type: ComponentType = ComponentType.LIBR
                  publisher: Optional[str] = None, group: Optional[str] = None, version: Optional[str] = None,
                  description: Optional[str] = None, scope: Optional[ComponentScope] = None,
                  hashes: Optional[List[HashType]] = None, licenses: Optional[List[LicenseChoice]] = None,
-                 copyright: Optional[str] = None, purl: Optional[PackageURL] = None,
+                 copyright: Optional[str] = None, purl: Optional[PackageURL] = None, cpe: Optional[str] = None,
```
Comment (jkowalleck) [inline cyclonedx/model/component.py:106]: :x: 

adding a new parameter not at the end of a function/method definition is considered a  breaking change,
because a call via non-keyword-arguments (`foo(1,2, 3)` instead of `foo(1, bar=2, baz=3)`) would require changing of order of parameters downstream.

Therefore, please add new parameters to the end.

@madpah we could enforce that some parameters must be keyword-arguments (`kwargs`) by the following [PEP3102](https://www.python.org/dev/peps/pep-3102/). Enforcing keyword-arguments would be considered a breaking change, IMHO.

**Review by jkowalleck — CHANGES_REQUESTED** [review]

File: cyclonedx/model/component.py [hunk 2]
```diff
@@ -491,7 +506,7 @@ def __eq__(self, other: object) -> bool:
     def __hash__(self) -> int:
         return hash((
             self.author, self.bom_ref, self.copyright, self.description, str(self.external_references), self.group,
-            str(self.hashes), str(self.licenses), self.mime_type, self.name, self.properties, self.publisher, self.purl,
+            str(self.hashes), str(self.licenses), self.mime_type, self.name, self.properties, self.publisher, self.purl, self.cpe,
```
Comment (jkowalleck) [inline cyclonedx/model/component.py:509]: :+1:

File: cyclonedx/model/component.py [hunk 3]
```diff
@@ -391,6 +392,20 @@ def purl(self) -> Optional[PackageURL]:
     @purl.setter
     def purl(self, purl: Optional[PackageURL]) -> None:
         self._purl = purl
+        
+    @property
+    def cpe(self) -> Optional[str]:
+        """
+        Specifies a well-formed CPE name that conforms to the CPE 2.2 or 2.3 specification.
```
Comment (jkowalleck) [inline cyclonedx/model/component.py:399]: :+1:

Comment (jkowalleck) [reply cyclonedx/model/component.py:399]: this is exactly the text from the [Spec](https://github.com/CycloneDX/specification/blob/master/schema/bom-1.4.xsd)
could you add the docs reference from the spec, too? 

```text
See https://nvd.nist.gov/products/cpe
```

**Comment by jkowalleck** [comment]
Thank you for the PullRequest, @jblu42.
This looks like a very good idea.

Aside from the PR comments, there are a couple of other hygiene points to look at:

1. You need to sign your commits - please check [CONTRIBUTING](https://github.com/CycloneDX/cyclonedx-python-lib/blob/main/CONTRIBUTING.md)
1. Few code stlye analysis violations have been reported
1. JSON output is missing - XML output was done
1. add tests for the following scenario
  * CPE is set to `None`  - test setter/getter/constructor
  * CPE is set to some non-`None` value - test setter/getter/constructor
  * test JSON output with a non-`None` value
  * test XML output with a non-`None` value

do you have any additional remarks, @madpah

File: cyclonedx/model/component.py (hunk 1 above)
Comment (jkowalleck) [reply cyclonedx/model/component.py]: closed via f2d47b6

File: cyclonedx/model/component.py (hunk 3 above)
Comment (jkowalleck) [reply cyclonedx/model/component.py]: was fixed

**Comment by jblu42** [comment]
Sorry forgot some fixes in the last commit, now added.

I think for JSON output there are no changes required as it uses the JSON schema as reference and the CPE is included in there. At least in my tests the JSON output is working.

Are the tests required? There are no tests for the other optional attributes, but if you like I can add some, just need to look how the tests are defined.

File: cyclonedx/model/component.py [hunk 4]
```diff
@@ -107,7 +107,7 @@ def __init__(self, name: str, component_type: ComponentType = ComponentType.LIBR
                  external_references: Optional[List[ExternalReference]] = None,
                  properties: Optional[List[Property]] = None, release_notes: Optional[ReleaseNotes] = None,
                  # Deprecated parameters kept for backwards compatibility
-                 namespace: Optional[str] = None, license_str: Optional[str] = None
+                 namespace: Optional[str] = None, license_str: Optional[str] = None, cpe: Optional[str] = None
```
Comment (jkowalleck) [inline cyclonedx/model/component.py:110]: @madpah the depricated parametes on the bottom ... this makes it hard to add additional non-depricated parametes after them.
how do you suggest to add new parameters? should they be added before the depricateds? 
we could use  [PEP 3102](https://www.python.org/dev/peps/pep-3102/) to force them as keyword-arguments.

Comment (jkowalleck) [inline cyclonedx/model/component.py:110]: @jblu42  i see the confusion. 
please put the new parameter in a new line above the `# Deprecated parameters kept for backwards compatibility` 
just like done here: https://github.com/CycloneDX/cyclonedx-python-lib/pull/137/files#diff-b1cd177706b0ed9acbe48de561ac6f96186fc1a291b00e757b30b514fd3acadaR110

**Comment by jkowalleck** [comment]
> Sorry forgot some fixes in the last commit, now added.
> 
> I think for JSON output there are no changes required as it uses the JSON schema as reference and the CPE is included in there. At least in my tests the JSON output is working.
> 
> Are the tests required? There are no tests for the other optional attributes, but if you like I can add some, just need to look how the tests are defined.

If you can, please add  a test for the JSON and for the XML output, that includes a non-`None` CPE value. that should be good enough for now.
If you think additional tests would support the feature you provide, feel free to add them, too.

**Comment by jkowalleck** [comment]
@jblu42 ,
Could you get a couple of formals sorted out? 
* some commits are not signed. See the "DCO" test results for details.
* some test are failing. see the test results for details.

**Comment by jblu42** [comment]
> 
Strange, I did run it through poetry tests before committing, will have a look at it again. The problem with the sign off is my first commit, who was not signed off, do you know how to correct that, I am not using git so often?

**Comment by jkowalleck** [comment]
> Strange, I did run it through poetry tests before committing, will have a look at it again. The problem with the sign off is my first commit, who was not signed off, do you know how to correct that, I am not using git so often?

Hello @jblu42 ,
By following the steps DCO checker described in the "details" link in the test summary you should be all good.
See for example: https://github.com/CycloneDX/cyclonedx-python-lib/pull/138/checks?check_run_id=4893513908

**Review by jkowalleck — APPROVED** [review]

File: cyclonedx/model/component.py (hunk 4 above)
Comment (madpah) [reply cyclonedx/model/component.py]: @jkowalleck - I would have instinctively added `cpe` just before the deprecated parameters.

However, PEP-3102 looks to be of interest here as you point out. I assume you are suggesting the function signature be changed to something like:

```python
 ...
properties: Optional[List[Property]] = None, release_notes: Optional[ReleaseNotes] = None,
cpe: Optional[str] = None, *,
# Deprecated parameters kept for backwards compatibility
namespace: Optional[str] = None, license_str: Optional[str] = None
...
```

File: tests/fixtures/bom_v1.2_setuptools_with_cpe.json [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  },
+  "components": [
+    {
+      "type": "library",
+      "bom-ref": "pkg:pypi/setuptools@50.3.2?extension=tar.gz",
+      "author": "Test Author",
+      "name": "setuptools",
+      "version": "50.3.2",
+      "purl": "pkg:pypi/setuptools@50.3.2?extension=tar.gz",
+      "cpe": "cpe:2.3:a:python:setuptools:50.3.2:*:*:*:*:*:*:*"
```
Comment (madpah) [inline tests/fixtures/bom_v1.2_setuptools_with_cpe.json:25]: Minor point: `cpe` comes prior to `purl` in the specification (align with all other test cases).

File: tests/fixtures/bom_v1.3_setuptools_with_cpe.json [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    ]
+  },
+  "components": [
+    {
+      "type": "library",
+      "name": "setuptools",
+      "version": "50.3.2",
+      "cpe": "cpe:2.3:a:python:setuptools:50.3.2:*:*:*:*:*:*:*",
+      "purl": "pkg:pypi/setuptools@50.3.2?extension=tar.gz",
+      "bom-ref": "pkg:pypi/setuptools@50.3.2?extension=tar.gz",
```
Comment (madpah) [inline tests/fixtures/bom_v1.3_setuptools_with_cpe.json:24]: Minor point: `bom-ref` comes before `name` in specification (align with other test cases).

File: tests/fixtures/bom_v1.4_setuptools_with_cpe.json [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    ]
+  },
+  "components": [
+    {
+      "type": "library",
+      "name": "setuptools",
+      "version": "50.3.2",
+	  "cpe": "cpe:2.3:a:python:setuptools:50.3.2:*:*:*:*:*:*:*",
+      "purl": "pkg:pypi/setuptools@50.3.2?extension=tar.gz",
+      "bom-ref": "pkg:pypi/setuptools@50.3.2?extension=tar.gz"
```
Comment (madpah) [inline tests/fixtures/bom_v1.4_setuptools_with_cpe.json:58]: Minor point: `bom-ref` comes before `name` in specification (align with other test cases).

**Review by madpah — CHANGES_REQUESTED** [review]
This is looking great. Couple of very minor things in tests.

FYI @jkowalleck

File: tests/fixtures/bom_v1.4_setuptools_with_cpe.json (hunk 7 above)
Comment (jblu42) [reply tests/fixtures/bom_v1.4_setuptools_with_cpe.json:58]: @madpah @jkowalleck I will fix the indentation and the cpe position. 

I can also move the bom-ref before name, but these files were copies from the other test cases, where bom-ref is also at this position (currently after name). Either I do not change these, change them for my new tests or change them for all tests. Just let me know what you prefer.

Comment (madpah) [reply tests/fixtures/bom_v1.4_setuptools_with_cpe.json:58]: Fair comment @jblu42 - leave as is then. We can adjust all together at some point.

**Comment by madpah** [comment]
Thanks again @jblu42 - we'll just wait for @jkowalleck to come back regarding positioning of the new `cpe` parameter - but this is looking great and we really really appreciate the contribution.

**Comment by jkowalleck** [comment]
PR looks solid.  approval from my side.

File: cyclonedx/model/component.py (hunk 4 above)
Comment (madpah) [reply cyclonedx/model/component.py]: Raised https://github.com/CycloneDX/cyclonedx-python-lib/issues/143 to cover this later.

**Review by madpah — APPROVED** [review]
